irohad/ordering/impl/on_demand_ordering_service_impl.cpp

```diff
@@ -47,23 +48,33 @@ void OnDemandOrderingServiceImpl::onCollaborationOutcome(
 }
 
 void OnDemandOrderingServiceImpl::onBatches(CollectionType batches) {
-  for (auto &batch : batches) {
-    if (not batchAlreadyProcessed(*batch)) {
-      insertBatchToCache(batch);
-    }
-  }
+  for (auto &batch : batches)
+    if (not batchAlreadyProcessed(*batch))
+      if (!insertBatchToCache(batch))
+        break;
+
   log_->info("onBatches => collection size = {}", batches.size());
 }
 
 // ---------------------------------| Private |---------------------------------
-void OnDemandOrderingServiceImpl::insertBatchToCache(
+bool OnDemandOrderingServiceImpl::insertBatchToCache(
     std::shared_ptr<shared_model::interface::TransactionBatch> const &batch) {
   std::lock_guard<std::shared_timed_mutex> lock(batches_cache_cs_);
+  if (cached_txs_size_ + boost::size(batch->transactions())
+      > (transaction_limit_ * 1000ull)) {
+    log_->warn("Transactions cache overload. Batch {} skipped.", *batch);
+    return false;
+  }
   if (used_batches_cache_.find(batch) == used_batches_cache_.end()) {
     batches_cache_.insert(batch);
+    cached_txs_size_ += boost::size(batch->transactions());
```

Comment:
I would suggest to use `ByteSizeLong` on protobuf instead, because transactions can be large which would still cause issues.